dynamiqs/smesolve/sme_solver.py

```diff
@@ -0,0 +1,110 @@
+        super().__init__(H, y0, t_save, t_save_y, exp_ops, options, jump_ops=jump_ops)
+
+        self.meas_ops = meas_ops  # (len(meas_ops), n, n)
+        self.t_save_meas = t_save_meas  # (len(t_save_meas))
+        self.etas = etas  # (len(meas_ops))
+        self.generator = generator
+
+        self.save_meas_counter = 0
+
+        # initialize additional save tensors
+        batch_sizes = self.y0.shape[:-2]
+
+        self.meas_shape = (*batch_sizes, len(self.meas_ops))
+
+        # meas_save: (..., len(meas_ops), len(t_save_meas))
+        if len(t_save_meas) > 0:
+            self.meas_save = torch.zeros(
+                *self.meas_shape,
+                len(self.t_save_meas) - 1,
+                dtype=self.dtype_real,
+                device=self.device,
+            )
+        else:
+            self.meas_save = None
+
+        # tensor to hold the sum of measurement results on a time bin
+        self.bin_meas = torch.zeros(self.meas_shape)  # (..., len(etas))
+
+    def _save(self, t: float, y: Tensor):
+        if t in self.t_save_y:
+            super()._save(t, y)
+        if t in self.t_save_meas:
+            self.save_meas()
+
+    def save_meas(self):
+        if self.save_meas_counter != 0:
+            t_bin = (
+                self.t_save_meas[self.save_meas_counter]
+                - self.t_save_meas[self.save_meas_counter - 1]
+            )
+            self.meas_save[..., self.save_meas_counter - 1] = self.bin_meas / t_bin
+            self.bin_meas = torch.zeros_like(self.bin_meas)
+        self.save_meas_counter += 1
+
+    def sample_wiener(self, dt: float) -> Tensor:
+        # -> (b_H, b_rho, ntrajs)
+        return torch.normal(
+            torch.zeros(*self.meas_shape), sqrt(dt), generator=self.generator
+        ).to(dtype=self.dtype_real)
+
+    @cache
+    def Lp(self, rho: Tensor) -> Tensor:
+        # rho: (b_H, b_rho, ntrajs, n, n) -> (b_H, b_rho, ntrajs, n, n)
+        # L @ rho + rho @ Ldag
+        L_rho = torch.einsum('bij,...jk->...bik', self.meas_ops, rho)
+        return L_rho + L_rho.mH
+
+    def diff_backaction(self, rho: Tensor, dw: Tensor) -> Tensor:
+        # Compute the measurement backaction term of the diffusive SME.
+        # $$ \sum_k \sqrt{\eta_k} \mathcal{M}[L_k](\rho) dW_k $$
+        # where
+        # $$ \mathcal{M}[L](\rho) = L \rho + \rho L^\dag -
+        # \mathrm{Tr}\left[(L + L^\dag) \rho\right] \rho $$
+
+        # rho: (b_H, b_rho, ntrajs, n, n) -> (b_H, b_rho, ntrajs, n, n)
+
+        # L @ rho + rho @ Ldag
+        Lp_rho = self.Lp(rho)  # (..., b, n, n)
+
+        # L @ rho + rho @ Ldag - Tr(L @ rho + rho @  Ldag) rho
+        tmp = Lp_rho - torch.einsum('...bii,...jk->...bjk', Lp_rho, rho)
+        # tmp: (..., b, n, n)
+
+        # sum sqrt(eta) * dw * [L @ rho + rho @ Ldag - Tr(L @ rho + rho @ Ldag) rho]
+        fact = (self.etas.sqrt() * dw).to(self.dtype)
+        return torch.einsum('...b,...bij->...ij', fact, tmp)
```

Comment:
No need to use einsum for this type of simple operation. It's probably more costly due to the checks it runs. Also I don't think conversion to complex is useful as long as the precision is the same.
```suggestion
        fact = self.etas.sqrt() * dw
        return (fact[..., None, None] * tmp).sum(-3)
```

Reply:
Also I don't like the naming `fact`, I'm reading "factorial". Just use the full name instead, so `factor` or `prefactor` even.